### PR TITLE
Allow the user time-consuming cancel implementations

### PR DESCRIPTION
**Pull request by Timple**
by setting the flag after function succeeds

Solves #171

### CODE REVIEW

File: mbf_abstract_nav/src/abstract_controller_execution.cpp [hunk 1]
```diff
@@ -233,14 +233,14 @@ bool AbstractControllerExecution::reachedGoalCheck()
 
 bool AbstractControllerExecution::cancel()
 {
-  cancel_ = true;
   // returns false if cancel is not implemented or rejected by the recovery behavior (will run until completion)
   if(!controller_->cancel())
   {
     ROS_WARN_STREAM("Cancel controlling failed or is not supported by the plugin. "
                         << "Wait until the current control cycle finished!");
     return false;
```
Comment (corot) [inline mbf_abstract_nav/src/abstract_controller_execution.cpp]: We need the `cancel_ = true;` also here. Can you get the result of cancel() and return it after setting cancel_ to true?

```
bool AbstractControllerExecution::cancel()
{
  // returns false if cancel is not implemented or rejected by the recovery behavior (will run until completion)
  bool ctrl_cancelled = controller_->cancel();
  if(!ctrl_cancelled)
  {
    ROS_WARN_STREAM("Cancel controlling failed or is not supported by the plugin. "
                        << "Wait until the current control cycle finished!");
  }
  cancel_ = true;
  return ctrl_cancelled;
}
```

**Review by corot — CHANGES_REQUESTED** [review]

File: mbf_abstract_nav/src/abstract_controller_execution.cpp (hunk 1 above)
Comment (Timple) [reply mbf_abstract_nav/src/abstract_controller_execution.cpp]: Ah right, this was before my first coffee.
Something like this?
```
bool AbstractControllerExecution::cancel()
{
  // returns false if cancel is not implemented or rejected by the recovery behavior (will run until completion)
  bool can_cancel = controller_->cancel();
  ROS_WARN_STREAM_COND(!can_cancel, "Cancel controlling failed or is not supported by the plugin. "
                                        << "Wait until the current control cycle finished!");
  cancel_ = true;
  return can_cancel;
}
```

**Review by corot — APPROVED** [review]

**Comment by corot** [comment]
lgtm, will test during the weekend

**Comment by spuetz** [comment]
This is fine, logically it just moves the setting of the flag down, which will - as discussed - finish the cancel method first, and if the cancel method is not implemented, the controller execution will cancel the action after it returned from computeVel...

**Comment by Timple** [comment]
Would this feature be enough to justify a new debian release? :innocent:

**Comment by spuetz** [comment]
Sure, I will merge the other two open PR at the weekend and make a new release. ;) cheers.

**Comment by Timple** [comment]
ping :)

**Pull request merged**